### PR TITLE
Feat: Allow writeable & nonvisible fields as default sort

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

This work allows filtering and sorting by fields that are not visible but writtable in the Admin API, some example of such fields:

- createdBy
- updatedBy
- strapi_stage (review workflow)
- strapi_assignee (review workflow)

We have the requirement to filter and sort by those fields. Because of that it makes sense to allow setting those fields as a default sort also.

This is a workaround to make it work in V4, but we will be refactoring this work in V5. This PR allows those fields to be filtered and sorted, and so we will need to hardcode those fields in the frontend side.

### CODE REVIEW

**Comment by Marc-Roig** [comment]
@alexandrebodin  done!  🚀

File: api-tests/core/content-manager/fields/non-visible.test.api.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * == Test Suite Overview ==
+ *
+ * N°   Description
+ * -------------------------------------------
+ * 1.  Filters by non visible field (successfully)
+ * 2.  Filters by created_by (successfully)
+ * 3.  Filters by updated_by (successfully)
+ */
+
+describe('Test type decimal', () => {
```
Comment (alexandrebodin) [inline api-tests/core/content-manager/fields/non-visible.test.api.js:53]: copy paste I imagine :)

**Review by alexandrebodin — COMMENTED** [review]
LGTM just one typo :)

File: api-tests/core/content-manager/fields/non-visible.test.api.js (hunk 1 above)
Comment (Marc-Roig) [reply api-tests/core/content-manager/fields/non-visible.test.api.js:53]: oops yes ! On my way!

**Comment by Marc-Roig** [comment]
I will schedule this for the 4.12.0  release

File: packages/core/content-manager/server/controllers/validation/model-configuration.js [hunk 2]
```diff
@@ -32,7 +43,10 @@ const createSettingsSchema = (schema) => {
       // should be reset when the type changes
       mainField: yup.string().oneOf(validAttributes.concat('id')).default('id'),
       // should be reset when the type changes
-      defaultSortBy: yup.string().oneOf(validAttributes.concat('id')).default('id'),
+      defaultSortBy: yup
+        .string()
+        .oneOf(validAttributes.concat(['id', ...nonVisibleWritableAttributes]))
```
Comment (alexandrebodin) [inline packages/core/content-manager/server/controllers/validation/model-configuration.js:48]: I was wondering should we allow sorting by created by & updated by also (maybe the other PR should do that 🤔 )

File: packages/core/content-manager/server/controllers/validation/model-configuration.js [hunk 3]
```diff
@@ -22,6 +26,13 @@ module.exports = (schema, opts = {}) =>
 const createSettingsSchema = (schema) => {
   const validAttributes = Object.keys(schema.attributes).filter((key) => isListable(schema, key));
 
+  // TODO V5: Refactor non visible fields to be a part of content-manager schema
+  const model = strapi.getModel(schema.uid);
+  const nonVisibleWritableAttributes = intersection(
```
Comment (alexandrebodin) [inline packages/core/content-manager/server/controllers/validation/model-configuration.js:31]: Just noticed localizations end up in that list 🤔 I'm not sure that makes sense but not sure we can set them as non writable. Do you mind checking that ?

Comment (Marc-Roig) [reply packages/core/content-manager/server/controllers/validation/model-configuration.js:31]: 👍 I will check it. Do you think it's a problem if we allow to sort by localizations & locale?

Comment (Marc-Roig) [reply packages/core/content-manager/server/controllers/validation/model-configuration.js:31]: Sorting by those fields is already allowed as of today.

Comment (alexandrebodin) [reply packages/core/content-manager/server/controllers/validation/model-configuration.js:31]: Hum 🤔 by locale yes localizations I'm surprised but if that's the case then there is no reason you can do it in the API and not the Admin even if that would make much sense to filter by localizations anyway

**Review by alexandrebodin — COMMENTED** [review]
Tested the API endpoints and could test filtering on multiple different fields including `strapi_reviewWorkflows_stage` correctly. 

The only thing I'm not a fan with merge this now is that those fields don't appear in the UI so I get empty inputs in the configure the view if I defined them in the defaultSortBy. Should we merge this with some frontend ?

**Comment by Marc-Roig** [comment]
@alexandrebodin what do you think would be the appropiate change here :
- make all the sortable fields , including non visisble oness, available in the defaultSort list input
- If the default sort is set to a field that its not in the list, make it appear in the list.

cc @gu-stav

File: packages/core/content-manager/server/controllers/validation/model-configuration.js (hunk 2 above)
Comment (Marc-Roig) [reply packages/core/content-manager/server/controllers/validation/model-configuration.js]: updated to allow also creator fields [`f6543f7` (#17205)](https://github.com/strapi/strapi/pull/17205/commits/f6543f7003bdeef8bffd518a9ddfbd8f96e300ae)

**Comment by alexandrebodin** [comment]
> @alexandrebodin what do you think would be the appropiate change here :
> 
> * make all the sortable fields , including non visisble oness, available in the defaultSort list input
> * If the default sort is set to a field that its not in the list, make it appear in the list.
> 
> cc @gu-stav

I would say we should ensure both lists (FE/BE) are the same (displayed & validated against)

**Comment by Marc-Roig** [comment]
@alexandrebodin I think that could be tricky today, because forntend displays all the visible fields that the backend returns.
We could do a bigger refactor but those fields are based on a [service of the code](https://github.com/strapi/strapi/blob/main/packages/core/content-manager/server/services/data-mapper.js#L40:L53) that is heavely used in more places, and I am afraid changing that part could have more impact.

Our initial hope was to have another definition of visible fields, but turned out that will not be possible until v5.

Do you think we should push for ensuring this lists are consistent at this moment? I have the feeling is quite a big amount of work when we will also be refactoring it on v5 soon.

**Comment by alexandrebodin** [comment]
> @alexandrebodin I think that could be tricky today, because forntend displays all the visible fields that the backend returns. We could do a bigger refactor but those fields are based on a [service of the code](https://github.com/strapi/strapi/blob/main/packages/core/content-manager/server/services/data-mapper.js?rgh-link-date=2023-07-17T16%3A31%3A58Z#L40:L53) that is heavely used in more places, and I am afraid changing that part could have more impact.
> 
> Our initial hope was to have another definition of visible fields, but turned out that will not be possible until v5.
> 
> Do you think we should push for ensuring this lists are consistent at this moment? I have the feeling is quite a big amount of work when we will also be refactoring it on v5 soon.

Do we have a choice ? we will need to give the right info to the frontend if we want to allow filtering on those attributes they need to appear in the UI right? unless you just need it for the API and in that case I would prevent being able to set the as defaultSortBy probably :) 

For how we could do it;  I would probably go with an other approach knowing we can refactor it in v5. Something like adding this list of fields as a new info in the Schema the API returns. This would avoid all the side effects I think 🤔

**Comment by Marc-Roig** [comment]
@alexandrebodin I think at the moment is a bit tricky, because frontend doesn't yet filter by relations. And it's something we will do in the RW work.

Our original idea was to add those fields in the frontend (until we work on this on v5), handle relational sorting & filtering and showing the strapi stage field in the List View table like we show now.

image

Just in case we would want to return those fields in the api, I imagine we would be returning that in the /configuration endpoint. Does it really solve the issue? I think it would extend more in time the problem and adding more complexity in code that will have to be removed in months.

My guess is, for a content type settings you propose:

```js
{
  "uid": "api::category.category",
  "settings": {
    "bulkable": true,
    "filterable": true,
    "searchable": true,
    "pageSize": 10,
    "mainField": "name",
    "defaultSortBy": "name",
    "defaultSortOrder": "ASC"
  },
  "metadatas": {
    ...
    "strapi_stage": {
      "edit": {
        "label": "name",
        "description": "",
        "placeholder": "",
        "visible": false,
        "editable": true
      },
      "list": {
        "label": "name",
        "searchable": true,
        "sortable": true
        "visible": true // <= make it visible here?
      }
    },
  }
}
```

**Review by jhoward1994 — APPROVED** [review]

**Pull request merged**